trace_decoder/src/lib.rs

```diff
@@ -1,92 +1,63 @@
-//! Consider all TODOs to be tracked under that issue.
-//! </div>
+//! An _Ethereum Node_[^1] executes _transactions_ in _blocks_.
 //!
-//! Your neighborhood zk-ready [ethereum](https://github.com/0xPolygonZero/erigon)
-//! [node](https://github.com/0xPolygonHermez/cdk-erigon/) emits binary "witnesses"[^1].
+//! Execution mutates two key data structures:
+//! - [The state trie](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/#state-trie).
+//! - [The storage tries](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/#storage-trie).
 //!
-//! But [`plonky2`], your prover, wants [`GenerationInputs`].
+//! Ethereum nodes expose information about the transactions over RPC, e.g:
+//! - [The specific changes to the storage tries](TxnTrace::storage_written).
+//! - [Changes to account balance in the state trie](TxnTrace::balance).
 //!
-//! This library helps you get there.
+//! The tries and the transactions can be proved to be correct using a
+//! zero-knowledge prover like [`evm_arithmetization`],
+//! an Ethereum-y frontend for [`plonky2`].
 //!
-//! [^1]: A witness is an attestation of the state of the world, which can be
-//!       proven by a prover.
+//! **Prover perfomance is a high priority.**
 //!
-//! # Non-Goals
-//! - Performance - this won't be the bottleneck in any proving system.
-//! - Robustness - malicious or malformed input may crash this library.
+//! The aformentioned trie structures may have subtries _deferred_.
+//! That is, any node (and its children!) may be replaced by its hash,
+//! while maintaining provability of its contents:
+//! ```text
+//!     A               A
+//!    / \             / \
+//!   B   C     ->    H   C
+//!  / \   \               \
+//! D   E   F               F
+//! ```
+//! (where `H` is the hash of the `D/B\E` subtrie).
 //!
-//! TODO(0xaatif): https://github.com/0xPolygonZero/zk_evm/issues/275
-//!                refactor all the docs below
+//! The principle concern of this library is to step through the transactions,
+//! and reproduce the tries that the Ethereum node does,
+//! **while deferring all possible subtries to minimise prover load**,
+//! since prover performance is sensitive to the size of the trie.
 //!
-//! It might not be obvious why we need traces for each txn in order to generate
-//! proofs. While it's true that we could just run all the txns of a block in an
-//! EVM to generate the traces ourselves, there are a few major downsides:
-//! - The client is likely a full node and already has to run the txns in an EVM
-//!   anyways.
-//! - We want this protocol to be as agnostic as possible to the underlying
-//!   chain that we're generating proofs for, and running our own EVM would
-//!   likely cause us to loose this genericness.
+//! [^1]: In our stack, this is [a fork of erigon](https://github.com/0xPolygonZero/erigon),
+//!       which exposes more information over RPC.
```

Comment:
`Jerigon` is a temporary version to use the type 1 for. We are in the process of supporting cdk-erigon as default client, but also Polygon PoS, we have partners like IMX that will use the type1 with their Geth client, etc... so any specific mention here should probably be removed.

Reply:
Can this be deferred till that work is actually done?

Reply:
But this is already not true, as we already support vanilla Geth clients with the native tracer.